util/unifyfs-stage/src/unifyfs-stage.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2020, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ *
+ * Copyright 2020, UT-Battelle, LLC.
+ *
+ * LLNL-CODE-741539
+ * All rights reserved.
+ *
+ * This is the license for UnifyFS.
+ * For details, see https://github.com/LLNL/UnifyFS.
+ * Please read https://github.com/LLNL/UnifyFS/LICENSE for full license text.
+ */
+/* unifyfs-stage: this application is supposed to excuted by the unifyfs
+ * command line utility for:
+ * - stage in: moving files in pfs to unifyfs volume before user starts
+ *   application,
+ *   e.g., unifyfs start --stage-in=<manifest file>
+ * - stage out: moving files in the unifyfs volume to parallel file system
+ *   after user application completes,
+ *   e.g., unifyfs terminate --stage-out=<manifest file>
+ *
+ * Currently, we request users to pass the <manifest file> to specify target
+ * files to be transferred. The <manifest file> should list all target files
+ * and their destinations, line by line.
+ *
+ * This supports two transfer modes (although both are technically parallel):
+ *
+ * - serial: Each process will transfer a file. Data of a single file will
+ *   reside in a single compute node.
+ * - parallel (-p, --parallel): Each file will be split and transferred by all
+ *   processes. Data of a single file will be spread evenly across all
+ *   available compute nodes.
+ *
+ * TODO:
+ * Maybe later on, it would be better to have a size threshold. Based on the
+ * threshold, we can determine whether a file needs to transferred serially (if
+ * smaller than threshold), or parallelly.
+ */
+#include <config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <getopt.h>
+#include <errno.h>
+#include <limits.h>
+#include <libgen.h>
+#include <mpi.h>
+
+#include "unifyfs_const.h"
+#include "unifyfs-stage.h"
+
+int rank;
+int total_ranks;
+int verbose;
+
+static int debug;
+static int checksum;
+static int mode;
+static char* manifest_file;
+static char* mountpoint = "/unifyfs";
+static char* share_dir;
+
+static unifyfs_stage_t _ctx;
+
+/**
+ * @brief create a status file for unifyfs cli to get niotified with the data
+ * stage result
+ *
+ * @param status 0 indicates success
+ *
+ * @return 0 on success, errno otherwise
+ */
+static int create_status_file(int status)
+{
+    char filename[PATH_MAX] = { 0, };
+    FILE* fp = NULL;
+    const char* msg = status ? "fail" : "success";
+
+    sprintf(filename, "%s/%s", share_dir, UNIFYFS_STAGE_STATUS_FILENAME);
```

Comment:
It's safer to use `scnprintf()`

Reply:
Good point.  Changed in upcoming patch.

Reply:
If someone has any ideas as to what file to #include to get this function prototype, that would be helpful.  I've been looking for 20 minutes and can't find it.

Reply:
Never mind.  I found where the scnprintf kernel function is re-implemented in the unifyfs source tree.